src/xtgeo/surface/_regsurf_cube_window_v3.py

```diff
@@ -0,0 +1,373 @@
+        return cvalues, dvalues
+
+    logger.info("Original shape is %s", cvalues.shape)
+    cref = np.random.rand(cvalues.shape[0], cvalues.shape[1], ndiv * cvalues.shape[2])
+    dref = cref.copy()
+
+    num_points = cref.shape[-1]
+    # Create interpolation function for cube values
+    interp_func1 = interp1d(
+        np.arange(cvalues.shape[-1]),
+        cvalues,
+        axis=-1,
+        kind="linear",
+        fill_value="extrapolate",
+    )
+    interp_func2 = interp1d(
+        np.arange(dvalues.shape[-1]),
+        dvalues,
+        axis=-1,
+        kind="linear",
+        fill_value="extrapolate",
+    )
+    # Resample array2 to match the number of points in array1
+    cref = interp_func1(np.linspace(0, cvalues.shape[-1] - 1, num_points))
+    dref = interp_func2(np.linspace(0, dvalues.shape[-1] - 1, num_points))
+
+    logger.info("Resampling done, new shape is %s", cref.shape)
+    return cref, dref
+
+
+def _filter_cube_values_upper_lower(cvalues, dvalues, upper, lower):
+    """Filter the cube (cvalues) based on depth interval."""
+
+    nnans = np.count_nonzero(np.isnan(cvalues))
+    logger.info("Filter cube in depth... number of nans is %s", nnans)
+
+    upv = np.expand_dims(upper.values, axis=2)
+    lov = np.expand_dims(lower.values, axis=2)
+
+    cvalues[dvalues < upv] = np.nan
+    cvalues[dvalues > lov] = np.nan
+
+    nnans = np.count_nonzero(np.isnan(cvalues))
+    ndefi = np.count_nonzero(~np.isnan(cvalues))
+    logger.info("Filter cube in depth done, updated number of nans is %s", nnans)
+    logger.info("Filter cube in depth done, remaining is %s", ndefi)
+    return cvalues
+
+
+def _expand_attributes(attribute: Union[str, list]) -> list:
+    """The 'attribute' may be a name, 'all', or a list of attributes"""
+    useattrs = None
+    if isinstance(attribute, str):
+        if attribute == "all":
+            useattrs = ALLATTRS
+        else:
+            useattrs = [attribute]
+    else:
+        useattrs = attribute
+
+    if not all(item in ALLATTRS for item in useattrs):
+        raise ValueError(
+            f"One or more values are not a valid, input list is {useattrs}, "
+            f"allowed list is {ALLATTRS}"
+        )
+    return useattrs
+
+
+def _compute_stats(
+    cref: np.ndarray,
+    attr: str,
+    _regsurf_: xtgeo.RegularSurface,
+    upper: xtgeo.RegularSurface,
+    masksurf: xtgeo.RegularSurface,
+    sampling: str,
+    snapxy: bool,
+) -> xtgeo.RegularSurface:
+    """Compute the attribute and return the attribute map"""
+    logger.info("Compute stats...")
+
+    if attr == "mean":
+        values = np.nanmean(cref, axis=2)
+    elif attr == "var":
+        values = np.nanvar(cref, axis=2)
+    elif attr == "max":
+        values = np.nanmax(cref, axis=2)
+    elif attr == "min":
+        values = np.nanmin(cref, axis=2)
+    elif attr == "rms":
+        values = np.sqrt(np.nanmean(np.square(cref), axis=2))
```

Comment:
It seems like there's some abstraction potential here, where these statistical calculations might be better as attributes or methods of a wrapper class or dataclass. I don't have the full overview here yet, though

Reply:
Probably; we can consider in later refactoring